docs/guide/connection-settings.txt
=======================
URI Connection Settings
=======================

Optional connection settings are settings not covered by the :manual:`URI Connection String </reference/connection-string/>`. The following options are passed in the options parameter when you create a mongo client.

.. code-block:: js

   const { MongoClient } = require('mongodb');

   // Connection URL
   const url = 'mongodb://localhost:50000,localhost:50001';

   // create a client, passing in additional options
   const client = new MongoClient(url, {
     poolSize: 10, ssl: true
   });

   // Function to connect to the server and run your code
   async function run() {
     try {
       // Connect the client to the server
       await client.connect();
       console.log('Connected successfully to server');
     } finally {
       // Ensures that the client will close when you finish/error
       await client.close();
     }
   }

   // Runs your code
   run();

The table below shows all settings and what topology they affect.

.. list-table::
   :header-rows: 1

   * - Option
     - Affects
     - Type
     - Default
     - Description
   * - **useNewUrlParser**
     - Server, ReplicaSet, Mongos
     - boolean
     - false
     - Opt-in to the new MongoDB URI parser, which conforms to the common cross-driver URI parser specification. The new URI parser will become the only URI parser in the next major version of the driver.
   * - **useUnifiedTopology**
     - Server, ReplicaSet, Mongos
     - boolean
     - false
     - Opt-in to the new Server Discovery and Monitoring engine that is more stable and aligns with other drivers. This engine will become the only SDAM engine in the next major version of the driver.
   * - **validateOptions**
     - Server, ReplicaSet, Mongos
     - boolean
     - false
     - If true, the driver will error if it detects an option that it does not recognize. If false, the driver will merely warn.
   * - **poolSize**
     - Server, ReplicaSet, Mongos
     - integer
     - 5
     - Is an alias to :manual:`maxPoolSize </reference/connection-string/#urioption.maxPoolSize>`
   * - **minSize**
     - Server, ReplicaSet, Mongos
     - integer
     - 0
     - Is an alias to :manual:`maxPoolSize </reference/connection-string/#urioption.minPoolSize>`
   * - **sslValidate**
     - Server, ReplicaSet, Mongos
     - boolean
     - false
     - Validate mongod server certificate against ca. Is equivalent to :manual:`tlsInsecure </reference/connection-string/#urioption.tlsInsecure>`
   * - **family**
     - Server, ReplicaSet, Mongos
     - number
     - 0
     - Version of IP stack. Can be 4, 6, 0 or null (default). If 0 or null, will attempt to connect with IPv6, and will fall back to IPv4 on failure.
   * - **sslCA**
     - Server, ReplicaSet, Mongos
     - Array
     - null
     - Array of valid certificates either as Buffers or Strings
   * - **sslCert**
     - Server, ReplicaSet, Mongos
     - Buffer/String
     - null
     - String or buffer containing the certificate we wish to present
   * - **sslKey**
     - Server, ReplicaSet, Mongos
     - Buffer/String
     - null
     - String or buffer containing the certificate private key we wish to present
   * - **sslPass**
     - Server, ReplicaSet, Mongos
     - Buffer/String
     - null
     - String or buffer containing the certificate password
   * - **sslCRL**
     - Server, ReplicaSet, Mongos
     - Buffer[]/Buffer/string[]/string
     - null
     - Certificate Revocation Lists. See `tls.createSecureContext <https://nodejs.org/dist/latest-v10.x/docs/api/tls.html#tls_tls_createsecurecontext_options>`_
   * - **checkServerIdentity**
     - Server, ReplicaSet, Mongos
     - function/boolean
     - true
     - If a function, overrides built-in `tls.checkServerIdentity <https://nodejs.org/dist/latest-v10.x/docs/api/tls.html#tls_tls_checkserveridentity_hostname_cert>`_. See `tls.connect <https://nodejs.org/dist/latest-v10.x/docs/api/tls.html#tls_tls_connect_options_callback>`_. If ``false``, automatically verifies all certificates and servernames.
   * - **noDelay**
     - Server, ReplicaSet, Mongos
     - boolean
     - true
     - TCP Socket NoDelay option. See `socket.setNoDelay <https://nodejs.org/dist/latest-v10.x/docs/api/net.html#net_socket_setnodelay_nodelay>`_
   * - **keepAlive**
     - Server, ReplicaSet, Mongos
     - integer
     - 30000
     - Enables ``keepAlive`` on the TCP socket. See `the first parameter of socket.setKeepAlive <https://nodejs.org/dist/latest-v10.x/docs/api/net.html#net_socket_setkeepalive_enable_initialdelay>`_
   * - **keepAliveInitialDelay**
     - Server, ReplicaSet, Mongos
     - integer
     - 30000
     - The number of milliseconds to wait before initiating keepAlive on the TCP socket. See `the second parameter of socket.setKeepAlive <https://nodejs.org/dist/latest-v10.x/docs/api/net.html#net_socket_setkeepalive_enable_initialdelay>`_
   * - **connectTimeoutMS**
     - Server, ReplicaSet, Mongos
     - integer
     - 30000
     - TCP Connection timeout setting. The same as :manual:`connectTimeoutMS </reference/connection-string/#urioption.connectTimeoutMS>`
   * - **socketTimeoutMS**
     - Server, ReplicaSet, Mongos
     - integer
     - 360000
     - TCP Socket timeout setting. The same as :manual:`socketTimeoutMS </reference/connection-string/#urioption.socketTimeoutMS>`
   * - **acceptableLatencyMS**
     - Mongos
     - integer
     - 15
     - Deprecated alias to :manual:`localThresholdMS </reference/connection-string/#urioption.heartbeatFrequencyMS>`
   * - **w**
     - Server, ReplicaSet, Mongos
     - string, integer
     - null
     - The write concern.
   * - **wtimeout**
     - Server, ReplicaSet, Mongos
     - integer
     - null
     - Deprecated alias to :manual:`wtimeoutMS </reference/connection-string/#urioption.wtimeoutMS>`
   * - **j**
     - Server, ReplicaSet, Mongos
     - boolean
     - false
     - Deprecated alias to :manual:`journal <https://docs.mongodb.com/manual/reference/connection-string/#urioption.journal>`
   * - **forceServerObjectId**
     - Server, ReplicaSet, Mongos
     - boolean
     - false
     - Force server to assign _id values instead of driver.
   * - **serializeFunctions**
     - Server, ReplicaSet, Mongos
     - boolean
     - false
     - Serialize functions on any object.
   * - **ignoreUndefined**
     - Server, ReplicaSet, Mongos
     - boolean
     - false
     - Specify if the BSON serializer should ignore undefined fields.
   * - **raw**
     - Server, ReplicaSet, Mongos
     - boolean
     - false
     - Return document results as raw BSON buffers.
   * - **promoteLongs**
     - Server, ReplicaSet, Mongos
     - boolean
     - true
     - Promotes Long values to number if they fit inside the 53 bits resolution.
   * - **promoteBuffers**
     - Server, ReplicaSet, Mongos
     - boolean
     - false
     - Promotes Binary BSON values to native Node Buffers.
   * - **promoteValues**
     - Server, ReplicaSet, Mongos
     - boolean
     - true
     - Promotes BSON values to native types where possible, set to false to only receive wrapper types.
   * - **domainsEnabled**
     - Server, ReplicaSet, Mongos
     - boolean
     - false
     - Deprecated. Enable the wrapping of the callback in the current domain, disabled by default to avoid perf hit.
   * - **pkFactory**
     - Server, ReplicaSet, Mongos
     - object
     - null
     - A primary key factory object for generation of custom _id keys.
   * - **promiseLibrary**
     - Server, ReplicaSet, Mongos
     - object
     - null
     - A Promise library class the application wishes to use such as Bluebird, must be ES6 compatible.
   * - **loggerLevel**
     - Server, Replicaset, Mongos
     - string
     - null
     - Specify the log level used by the driver logger (error/warn/info/debug)
   * - **logger**
     - Server, Replicaset, Mongos
     - object
     - null
     - Specify a customer logger mechanism, can be used to log using your app level logger

The following options do not work with ``useUnifiedTopology: true`` and are consider deprecated:


.. list-table::
   :header-rows: 1

   * - Option
     - Affects
     - Type
     - Default
     - Description
   * - **autoReconnect**
     - Server
     - boolean
     - true
     - Enables autoReconnecting when using a standalone server in the legacy topology setup
   * - **reconnectTries**
     - Server
     - integer
     - 30
     - When ``autoReconnect`` is enabled, the server attempt to reconnect #times.
   * - **reconnectInterval**
     - Server
     - integer
     - 1000
     - When ``autoReconnect`` is enabled, the server will wait # milliseconds between retries.
   * - **haInterval**
     - ReplicaSet, Mongos
     - integer
     - 10000,5000
     - Time between each replicaset status check.
   * - **secondaryAcceptableLatencyMS**
     - ReplicaSet
     - integer
     - 15
     - Sets the range of servers to pick when using NEAREST (lowest ping ms + the latency fence, ex: range of 1 to (1 + 15) ms).
   * - **connectWithNoPrimary**
     - ReplicaSet
     - boolean
     - false
     - Sets if the driver should connect even if no primary is available.
   * - **bufferMaxEntries**
     - Server, ReplicaSet, Mongos
     - integer
     - -1
     - Sets a cap on how many operations the driver will buffer up before giving up on getting a working connection, default is -1 which is unlimited.



Ensure your connection string is valid for Replica Sets
=======================================================

The connection string passed to the driver must use the fully qualified host names for the servers as set in the replicaset config. Given the following configuration settings for your replicaset.

.. code-block:: js

   {
       _id: 'testSet',
       version: 1,
       protocolVersion: 1,
       members: [
           {
               _id: 1,
               host: 'server1:31000',
           },
           {
               _id: 2,
               host: 'server2:31001',
           },
           {
               _id: 3,
               host: 'server3:31002',
           }
       ]
   }

You must ensure ``server1``, ``server2`` and ``server3`` are resolvable from the driver for the Replicaset discovery and failover to work correctly.


docs/reference/content/reference/connecting/connection-settings.md
+++
date = "2015-03-19T12:53:30-04:00"
title = "Connection Settings"
[menu.main]
  parent = "Connection Options"
  identifier = "Connection Settings"
  weight = 40
  pre = "<i class='fa'></i>"
+++

# URI Connection Settings

Optional connection settings are settings not covered by the [URI Connection String ](https://docs.mongodb.org/manual/reference/connection-string/). The following options are passed in the options parameter when you create a mongo client.

```js
const MongoClient = require('mongodb').MongoClient;
const assert = require('assert');

// Connection URL
const url = 'mongodb://localhost:50000,localhost:50001';
// Database Name
const dbName = 'myproject';

// create a client, passing in additional options
const client = new MongoClient(url, {
  poolSize: 10, ssl: true
});

// Use connect method to connect to the server
client.connect(function(err) {
  assert.equal(null, err);
  console.log("Connected correctly to server");

  client.close();
});
```

The table below shows all settings and what topology they affect.

| Option | Affects | Type | Default | Description |
| :----------| :------------------ | :------ | :------ |:------------- |
| **poolSize** | Server, ReplicaSet, Mongos | integer | 5 | Set the maximum poolSize for each individual server or proxy connection.|
| **ssl** | Server, ReplicaSet, Mongos | boolean | false | Use ssl connection |
| **sslValidate** | Server, ReplicaSet, Mongos | boolean | false | Validate mongod server certificate against ca |
| **sslCA** | Server, ReplicaSet, Mongos | Array | null | Array of valid certificates either as Buffers or Strings |
| **sslCert** | Server, ReplicaSet, Mongos | Buffer/String | null | String or buffer containing the certificate we wish to present |
| **sslKey** | Server, ReplicaSet, Mongos | Buffer/String | null | String or buffer containing the certificate private key we wish to present |
| **sslPass** | Server, ReplicaSet, Mongos | Buffer/String | null | String or buffer containing the certificate password |
| **autoReconnect** | Server | boolean | true | Reconnect on error. |
| **noDelay** | Server, ReplicaSet, Mongos | boolean | true | TCP Socket NoDelay option. |
| **keepAlive** | Server, ReplicaSet, Mongos | integer | 30000 | The number of milliseconds to wait before initiating keepAlive on the TCP socket. |
| **connectTimeoutMS** | Server, ReplicaSet, Mongos | integer | 30000 | TCP Connection timeout setting. |
| **socketTimeoutMS** | Server, ReplicaSet, Mongos | integer | 360000 | TCP Socket timeout setting. |
| **reconnectTries** | Server | integer | 30 | Server attempt to reconnect #times |
| **reconnectInterval** | Server | integer | 1000 | Server will wait # milliseconds between retries. |
| **ha** | ReplicaSet, Mongos | boolean | true | Turn on high availability monitoring. |
| **haInterval** | ReplicaSet, Mongos | integer | 10000,5000 | Time between each replicaset status check. |
| **replicaSet** | ReplicaSet | string | null | The name of the replicaset to connect to. |
| **secondaryAcceptableLatencyMS** | ReplicaSet | integer | 15 | Sets the range of servers to pick when using NEAREST (lowest ping ms + the latency fence, ex: range of 1 to (1 + 15) ms). |
| **acceptableLatencyMS** | Mongos | integer | 15 | Sets the range of servers to pick when using NEAREST (lowest ping ms + the latency fence, ex: range of 1 to (1 + 15) ms). |
| **connectWithNoPrimary** | ReplicaSet | boolean | false | Sets if the driver should connect even if no primary is available. |
| **authSource** | Server, ReplicaSet, Mongos | string | null |  If the database authentication is dependent on another databaseName. |
| **w** | Server, ReplicaSet, Mongos | string, integer| null |  The write concern. |
| **wtimeout** | Server, ReplicaSet, Mongos | integer | null |  The write concern timeout value. |
| **j** | Server, ReplicaSet, Mongos | boolean | false | Specify a journal write concern. |
| **forceServerObjectId** | Server, ReplicaSet, Mongos | boolean | false | Force server to assign _id values instead of driver. |
| **serializeFunctions** | Server, ReplicaSet, Mongos | boolean | false | Serialize functions on any object. |
| **ignoreUndefined** | Server, ReplicaSet, Mongos | boolean | false | Specify if the BSON serializer should ignore undefined fields. |
| **raw** | Server, ReplicaSet, Mongos | boolean | false | Return document results as raw BSON buffers. |
| **promoteLongs** | Server, ReplicaSet, Mongos | boolean | true | Promotes Long values to number if they fit inside the 53 bits resolution. |
| **promoteBuffers** | Server, ReplicaSet, Mongos | boolean | false | Promotes Binary BSON values to native Node Buffers. |
| **promoteValues** | Server, ReplicaSet, Mongos | boolean | true | Promotes BSON values to native types where possible, set to false to only receive wrapper types. |
| **domainsEnabled** | Server, ReplicaSet, Mongos | boolean | false | Enable the wrapping of the callback in the current domain, disabled by default to avoid perf hit. |
| **bufferMaxEntries** | Server, ReplicaSet, Mongos | integer | -1 | Sets a cap on how many operations the driver will buffer up before giving up on getting a working connection, default is -1 which is unlimited. |
| **readPreference** | Server, ReplicaSet, Mongos | object | null | The preferred read preference (ReadPreference.PRIMARY, ReadPreference.PRIMARY_PREFERRED, ReadPreference.SECONDARY, ReadPreference.SECONDARY_PREFERRED, ReadPreference.NEAREST). |
| **pkFactory** | Server, ReplicaSet, Mongos | object | null | A primary key factory object for generation of custom _id keys. |
| **promiseLibrary** | Server, ReplicaSet, Mongos | object | null | A Promise library class the application wishes to use such as Bluebird, must be ES6 compatible. |
| **readConcern** | Server, ReplicaSet, Mongos | object | null |  Specify a read concern for the collection. (only MongoDB 3.2 or higher supported). |
| **maxStalenessSeconds** | Replicaset | number | null | Specify a maxStalenessSeconds value for secondary reads, minimum is 90 seconds |
| **appname** | Server, Replicaset, Mongos | string | null | The name of the application that created this MongoClient instance. MongoDB 3.4 and newer will print this value in the server log upon establishing each connection. It is also recorded in the slow query log and profile collections. |
| **loggerLevel** | Server, Replicaset, Mongos | string | null | Specify the log level used by the driver logger (error/warn/info/debug) |
| **logger** | Server, Replicaset, Mongos | object | null | Specify a customer logger mechanism, can be used to log using your app level logger |
# Ensure your connection string is valid for Replica Sets

The connection string passed to the driver must use the fully qualified host names for the servers as set in the replicaset config. Given the following configuration settings for your replicaset.

```js
{
	"_id" : "testSet",
	"version" : 1,
	"protocolVersion" : 1,
	"members" : [
		{
			"_id" : 1,
			"host" : "server1:31000",
		},
		{
			"_id" : 2,
			"host" : "server2:31001",
		},
		{
			"_id" : 3,
			"host" : "server3:31002",
		}
	]
}
```

You must ensure `server1`, `server2` and `server3` are resolvable from the driver for the Replicaset discovery and failover to work correctly.


HISTORY
1.3.23 2013-12-13
-----------------
- Bumped bson package to 0.2.5 to work around resolution of include files issues for a lot of node.js installs

1.3.22 2013-12-12
-----------------
- Bumped bson package to 0.2.4 to work around 0.10.22 issues

1.3.21 2013-12-11
-----------------
- Fix for MongoClient connection timeout issue (NODE-97)

1.3.20 2013-12-11
-----------------
- All node exceptions will no longer be caught if on('error') is defined
- Added X509 auth support
- Pass through error messages from parseError instead of just text (Issue #1125)
- Close db connection on error (Issue #1128, https://github.com/benighted)
- Fixed documentation generation
- Added bufferMaxEntries to start failing if the buffer hits the specified number of entries
- Upgraded BSON parser to work with 0.12.X C++ API changes
- Support for OP_LOG_REPLAY flag (NODE-94)

1.3.19 2013-08-21
-----------------
- Correctly rethrowing errors after change from event emission to callbacks, compatibility with 0.10.X domains without breaking 0.8.X support.
- Small fix to return the entire findAndModify result as the third parameter (Issue #1068)
- No removal of "close" event handlers on server reconnect, emits "reconnect" event when reconnection happens. Reconnect Only applies for single server connections as of now as semantics for ReplSet and Mongos is not clear (Issue #1056)

1.3.18 2013-08-10
-----------------
- Fixed issue when throwing exceptions in MongoClient.connect/Db.open (Issue #1057)
- Fixed an issue where _events is not cleaned up correctly causing a slow steady memory leak.

1.3.17 2013-08-07
-----------------
- Ignore return commands that have no registered callback
- Made collection.count not use the db.command function
- Fix throw exception on ping command (Issue #1055)

1.3.16 2013-08-02
-----------------
- Fixes connection issue where lots of connections would happen if a server is in recovery mode during connection (Issue #1050, NODE-50, NODE-51)
- Bug in unlink mulit filename (Issue #1054)

1.3.15 2013-08-01
-----------------
- Memory leak issue due to node Issue #4390 where _events[id] is set to undefined instead of deleted leading to leaks in the Event Emitter over time

1.3.14 2013-08-01
-----------------
- Fixed issue with checkKeys where it would error on X.X

1.3.13 2013-07-31
-----------------
- Added override for checkKeys on insert/update (Warning will expose you to injection attacks) (Issue #1046)
- BSON size checking now done pre serialization (Issue #1037)
- Added isConnected returns false when no connection Pool exists (Issue #1043)
- Unified command handling to ensure same handling (Issue #1041, #1042)
- Correctly emit "open" and "fullsetup" across all Db's associated with Mongos, ReplSet or Server (Issue #1040)
- Correctly handles bug in authentication when attempting to connect to a recovering node in a replicaset.
- Correctly remove recovering servers from available servers in replicaset. Piggybacks on the ping command.
- Removed findAndModify chaining to be compliant with behavior in other official drivers and to fix a known mongos issue.
- Fixed issue with Kerberos authentication on Windows for re-authentication.
- Fixed Mongos failover behavior to correctly throw out old servers.
- Ensure stored queries/write ops are executed correctly after connection timeout
- Added promoteLongs option for to allow for overriding the promotion of Longs to Numbers and return the actual Long.

1.3.12 2013-07-19
-----------------
- Fixed issue where timeouts sometimes would behave wrongly (Issue #1032)
- Fixed bug with callback third parameter on some commands (Issue #1033)
- Fixed possible issue where killcursor command might leave hanging functions
- Fixed issue where Mongos was not correctly removing dead servers from the pool of eligable servers
- Throw error if dbName or collection name contains null character (at command level and at collection level)
- Updated bson parser to 0.2.1 with security fix and non-promotion of Long values to javascript Numbers (once a long always a long) 

1.3.11 2013-07-04
-----------------
- Fixed errors on geoNear and geoSearch (Issue #1024, https://github.com/ebensing)
- Add driver version to export (Issue #1021, https://github.com/aheckmann)
- Add text to readpreference obedient commands (Issue #1019)
- Drivers should check the query failure bit even on getmore response (Issue #1018)
- Map reduce has incorrect expectations of 'inline' value for 'out' option (Issue #1016, https://github.com/rcotter)
- Support SASL PLAIN authentication (Issue #1009)
- Ability to use different Service Name on the driver for Kerberos Authentication (Issue #1008)
- Remove unnecessary octal literal to allow the code to run in strict mode (Issue #1005, https://github.com/jamesallardice)
- Proper handling of recovering nodes (when they go into recovery and when they return from recovery, Issue #1027)

1.3.10 2013-06-17
-----------------
- Guard against possible undefined in server::canCheckoutWriter (Issue #992, https://github.com/willyaranda)
- Fixed some duplicate test names (Issue #993, https://github.com/kawanet)
- Introduced write and read concerns for GridFS (Issue #996)
- Fixed commands not correctly respecting Collection level read preference (Issue #995, #999)
- Fixed issue with pool size on replicaset connections (Issue #1000)
- Execute all query commands on master switch (Issue #1002, https://github.com/fogaztuc)

1.3.9 2013-06-05
----------------
- Fixed memory leak when findAndModify errors out on w>1 and chained callbacks not properly cleaned up.

1.3.8 2013-05-31
----------------
- Fixed issue with socket death on windows where it emits error event instead of close event (Issue #987)
- Emit authenticate event on db after authenticate method has finished on db instance (Issue #984)
- Allows creation of MongoClient and do new MongoClient().connect(..). Emits open event when connection correct allowing for apps to react on event.

1.3.7 2013-05-29
----------------

lib/mongodb/connection/repl_set/options.js
var PingStrategy = require('./strategies/ping_strategy').PingStrategy
  , StatisticsStrategy = require('./strategies/statistics_strategy').StatisticsStrategy
  , ReadPreference = require('../read_preference').ReadPreference;

var Options = function(options) {
  options = options || {};
  this._options = options;
  this.ha = options.ha || true;
  this.haInterval = options.haInterval || 2000;
  this.reconnectWait = options.reconnectWait || 1000;
  this.retries = options.retries || 30;
  this.rs_name = options.rs_name;
  this.socketOptions = options.socketOptions || {};
  this.readPreference = options.readPreference;
  this.readSecondary = options.read_secondary;
  this.poolSize = options.poolSize == null ? 5 : options.poolSize;
  this.strategy = options.strategy || 'ping';
  this.secondaryAcceptableLatencyMS = options.secondaryAcceptableLatencyMS || 15;
  this.connectArbiter = options.connectArbiter || false;
  this.connectWithNoPrimary = options.connectWithNoPrimary || false;
  this.logger = options.logger;
  this.ssl = options.ssl || false;
  this.sslValidate = options.sslValidate || false;
  this.sslCA = options.sslCA;
  this.sslCert = options.sslCert;
  this.sslKey = options.sslKey;
  this.sslPass = options.sslPass;
  this.emitOpen = options.emitOpen || true;
}

Options.prototype.init = function() {
  if(this.sslValidate && (!Array.isArray(this.sslCA) || this.sslCA.length == 0)) {
    throw new Error("The driver expects an Array of CA certificates in the sslCA parameter when enabling sslValidate");
  }  

  // Make sure strategy is one of the two allowed
  if(this.strategy != null && (this.strategy != 'ping' && this.strategy != 'statistical' && this.strategy != 'none')) 
      throw new Error("Only ping or statistical strategies allowed");    
  
  if(this.strategy == null) this.strategy = 'ping';
  
  // Set logger if strategy exists
  if(this.strategyInstance) this.strategyInstance.logger = this.logger;

  // Unpack read Preference
  var readPreference = this.readPreference;
  // Validate correctness of Read preferences
  if(readPreference != null) {
    if(readPreference != ReadPreference.PRIMARY && readPreference != ReadPreference.PRIMARY_PREFERRED
      && readPreference != ReadPreference.SECONDARY && readPreference != ReadPreference.SECONDARY_PREFERRED
      && readPreference != ReadPreference.NEAREST && typeof readPreference != 'object' && readPreference['_type'] != 'ReadPreference') {
      throw new Error("Illegal readPreference mode specified, " + JSON.stringify(readPreference));
    }

    this.readPreference = readPreference;
  } else {
    this.readPreference = null;
  } 

     // Ensure read_secondary is set correctly
  if(this.readSecondary != null)
    this.readSecondary = this.readPreference == ReadPreference.PRIMARY 
        || this.readPreference == false  
        || this.readPreference == null ? false : true;

  // Ensure correct slave set
  if(this.readSecondary) this.slaveOk = true;

  // Set up logger if any set
  this.logger = this.logger != null
    && (typeof this.logger.debug == 'function')
    && (typeof this.logger.error == 'function')
    && (typeof this.logger.debug == 'function')
      ? this.logger : {error:function(message, object) {}, log:function(message, object) {}, debug:function(message, object) {}};  

  // Connection timeout
  this.connectTimeoutMS = typeof this.socketOptions.connectTimeoutMS == 'number'
    ? this.socketOptions.connectTimeoutMS
    : 1000;

  // Socket connection timeout
  this.socketTimeoutMS = typeof this.socketOptions.socketTimeoutMS == 'number'
    ? this.socketOptions.socketTimeoutMS
    : 30000;
}

Options.prototype.decorateAndClean = function(servers, callBackStore) {
  var self = this;

  // var de duplicate list
  var uniqueServers = {};
  // De-duplicate any servers in the seed list
  for(var i = 0; i < servers.length; i++) {
    var server = servers[i];
    // If server does not exist set it
    if(uniqueServers[server.host + ":" + server.port] == null) {
      uniqueServers[server.host + ":" + server.port] = server;
    }
  }

  // Let's set the deduplicated list of servers
  var finalServers = [];
  // Add the servers
  for(var key in uniqueServers) {
    finalServers.push(uniqueServers[key]);
  }

  finalServers.forEach(function(server) {
    // Ensure no server has reconnect on
    server.options.auto_reconnect = false;
    // Set up ssl options
    server.ssl = self.ssl;
    server.sslValidate = self.sslValidate;
    server.sslCA = self.sslCA;
    server.sslCert = self.sslCert;
    server.sslKey = self.sslKey;
    server.sslPass = self.sslPass;
    server.poolSize = self.poolSize;
    // Set callback store
    server._callBackStore = callBackStore;
  });

  return finalServers;
}

exports.Options = Options;
